eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs

```diff
@@ -0,0 +1,805 @@
+    Just Refl <- sameNat (Proxy @(SizeHash (CC.ADDRHASH (Crypto era)))) (Proxy @28) =
+    let addr = decodeAddress28 stakeRef a b c d
+     in (addr, inject (fromCompact adaVal), SJust (decodeDataHash32 e f g h))
+viewTxOut (TxOut_AddrHash28_AdaOnly {}) = error "Impossible: Compacted and address or hash of non-standard size"
+viewTxOut (TxOut_AddrHash28_AdaOnly_DataHash32 {}) = error "Impossible: Compacted and address or hash of non-standard size"
+
+instance
+  ( Era era,
+    Show (Core.Value era),
+    Show (CompactForm (Core.Value era))
+  ) =>
+  Show (TxOut era)
+  where
+  show = show . viewTxOut
+
+deriving via InspectHeapNamed "TxOut" (TxOut era) instance NoThunks (TxOut era)
+
+pattern TxOut ::
+  forall era.
+  ( Era era,
+    Compactible (Core.Value era),
+    Val (Core.Value era),
+    HasCallStack
+  ) =>
+  Addr (Crypto era) ->
+  Core.Value era ->
+  StrictMaybe (DataHash (Crypto era)) ->
+  TxOut era
+pattern TxOut addr vl dh <-
+  (viewTxOut -> (addr, vl, dh))
+  where
+    TxOut (Addr network paymentCred stakeRef) vl SNothing
+      | StakeRefBase stakeCred <- stakeRef,
+        Just adaCompact <- getAdaOnly (Proxy @era) vl,
+        Just (Refl, a, b, c, d) <- encodeAddress28 network paymentCred =
+        TxOut_AddrHash28_AdaOnly stakeCred a b c d adaCompact
+    TxOut (Addr network paymentCred stakeRef) vl (SJust dh)
+      | StakeRefBase stakeCred <- stakeRef,
+        Just adaCompact <- getAdaOnly (Proxy @era) vl,
+        Just (Refl, a, b, c, d) <- encodeAddress28 network paymentCred,
+        Just (Refl, e, f, g, h) <- encodeDataHash32 dh =
+        TxOut_AddrHash28_AdaOnly_DataHash32 stakeCred a b c d adaCompact e f g h
+    TxOut addr vl mdh =
+      let v = fromMaybe (error "Illegal value in txout") $ toCompact vl
+          a = compactAddr addr
+       in case mdh of
+            SNothing -> TxOutCompact a v
+            SJust dh -> TxOutCompactDH a v dh
+
+{-# COMPLETE TxOut #-}
+
+-- ======================================
+
+type ScriptIntegrityHash crypto = SafeHash crypto EraIndependentScriptIntegrity
+
+data TxBodyRaw era = TxBodyRaw
+  { _inputs :: !(Set (TxIn (Crypto era))),
+    _collateral :: !(Set (TxIn (Crypto era))),
+    _collateralReturn :: !(StrictMaybe (TxOut era)),
+    _outputs :: !(StrictSeq (TxOut era)),
+    _certs :: !(StrictSeq (DCert (Crypto era))),
+    _wdrls :: !(Wdrl (Crypto era)),
+    _txfee :: !Coin,
+    _vldt :: !ValidityInterval,
+    _update :: !(StrictMaybe (Update era)),
+    _reqSignerHashes :: Set (KeyHash 'Witness (Crypto era)),
+    _mint :: !(Value (Crypto era)),
+    -- The spec makes it clear that the mint field is a
+    -- Cardano.Ledger.Mary.Value.Value, not a Core.Value.
+    -- Operations on the TxBody in the AlonzoEra depend upon this.
+    _scriptIntegrityHash :: !(StrictMaybe (ScriptIntegrityHash (Crypto era))),
+    _adHash :: !(StrictMaybe (AuxiliaryDataHash (Crypto era))),
+    _txnetworkid :: !(StrictMaybe Network)
+  }
+  deriving (Generic, Typeable)
```

Comment:
@JaredCorduan or @nc6 can you comment on this? I think we could do something like this instead?

```suggestion
data TxBodyRaw era = TxBodyRaw
  { txBodyAlonzo :: Alonzo.TxBodyRaw era
  , txBodyCollateralReturn :: !(StrictMaybe (TxOut era))
  }
```

This way we could reuse all the serialization/deserialization logic from alonzo. All this copy pasting is hurting my eyes. I will of course not stand in a way if that is how `era` parameterization was designed to be used.

Reply:
hmm. I like the idea of removing code copy when we can, but it's also weird to switch to this nested format now.

Reply:
We could have the datatype defined like this, and unpack in the patterns.